### PR TITLE
Update test documentation

### PR DESCRIPTION
**Pull request by namurphy**
Our test practices have been evolving and improving over the last several months, and this documentation aims to update our development guide to reflect those changes.  

I am not entirely familiar with all of the aspects of our testing, so suggestions and contributions from others would be much appreciated!

Cross-references: #153, #197, #198, #246, #249, #278, #291, #490

### CODE REVIEW

**Comment by StanczakDominik** [comment]
The changes look okay... I'll have to take a look at why the doc build is failing.

**Comment by StanczakDominik** [comment]
Well taken a look! That looks like it's going to work :)

**Comment by namurphy** [comment]
The main remaining task is to add some information about fixtures that describes what they are and how they work, and provide some straightforward, minimal examples.

File: docs/development/testing_guide.rst [hunk 1]
```diff
@@ -282,12 +517,40 @@ The coverage reports may be accessed by opening the newly generated
...
+
+To ignore a line of code in coverage tests, append it with
+``# coverage: ignore``.  If this comment is used on a line with a
+control flow structure (e.g., `if`, `for`, and `while`) that begins a
+block of code, then all lines in that block of code will be ignored.  In
+the following example, lines 3 and 4 will be ignored in coverage tests.
+
+.. code-block:: python
+  :linenos:
+  :emphasize-lines: 3,4
```
[the remark this answers is not in the document]
Comment (namurphy) [reply docs/development/testing_guide.rst]: Yeah, I had to look up how to do this.  Looks really useful!

**Comment by StanczakDominik** [comment]
Or we could document them here once we actually start to use them.

**Comment by namurphy** [comment]
@StanczakDominik - The fixtures section now contains a few sentences that refer to the pytest docs and say that we should use fixtures for more complex tests.

I've made all the changes I've been wanting to make, so this is ready for review.  Does anyone have any final suggestions?

I've come to appreciate squash merges for cases like this (thanks to @StanczakDominik and a few other people who brought this up in the last few months), and I plan to do that here.  There isn't anything in the individual commits that is worth saving, and bunching them together will lead to a cleaner history.  I generally prefer to do squash merges for my PRs myself, since that gives me a chance to write a detailed commit message with contextual information.

**Comment by StanczakDominik** [comment]
Oh right, now I remember what I wanted to say.

> I generally prefer to do squash merges for my PRs myself, since that gives me a chance to write a detailed commit message with contextual information.

If you ever get into a "PR is ready but waiting for review" situation and want to customize the eventual commit message, you could do a `git rebase -i master feature_branch` to squash the individual commits locally. I think that's going to preserve commit information on the github squash merge.

**Pull request merged**